accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,485 @@
+Debug.Assert(p1ReqMods[0].GetType() == typeof(Runtime.InteropServices.InAttribute));
+```
+
+```c#
+public class Holder
+{
+    // Array of function pointers
+    public unsafe delegate*<in int, out int, void>[] _field;
+}
+
+FieldInfo fieldInfo = typeof(Holder).GetField("_field");
+Debug.Assert(fieldInfo.FieldType.IsArray == true);
+
+Type fnType = fieldInfo.FieldType.ElementType;
+Debug.Assert(fnType.IsFunctionPointer);
+Type arg1Type = fnType.GetFunctionPointerParameterTypes()[0];
+// Custom modifiers not returned by default
+Debug.Assert(arg1Type.GetRequiredCustomModifiers().Length == 0);
+
+// TODO: discuss this
+// We want this Assert to pass in MetadataLoadContext and throw InvalidOperationException in CoreClr.
+// However, for nested types this implies a "calling context" is established once
+// GetFunctionPointerSignatureType() is called in order to populate ElementType.
+Debug.Assert(fieldInfo.GetFunctionPointerSignatureType().ElementType.GetRequiredCustomModifiers().Length == 1);
+// To avoid the "context" we can add helpers:
+Debug.Assert(fieldInfo.GetSignatureTypeOfArrayType().GetRequiredCustomModifiers().Length == 1);
+// but we'd also need similar helpers for pointers and generic parameters, and even those
+// would not work if nesting is more than one level deep such as an array of pointers to function pointers.
+```
+
+## System.Reflection.TypeDelegator
+No new APIs, but new overrides to support delegation.
+```diff
+namespace System.Reflection
+{
+    public class TypeDelegator : TypeInfo
+    {
++       public override Type GetFunctionPointerReturnType();
++       public override Type[] GetFunctionPointerParameterTypes();
++       public override bool IsFunctionPointer { get; }
++       public override bool IsUnmanagedFunctionPointer { get; }
++       public override Type[] GetFunctionPointerCallingConventions();
++       public override Type[] GetRequiredCustomModifiers();
++       public override Type[] GetOptionalCustomModifiers();
+    }
+}
+```
+
+## System.Reflection.Emit.ILGenerator
+See the existing [ILGenerator.EmitCalli()](https://docs.microsoft.com/dotnet/api/system.reflection.emit.ilgenerator.emitcalli#system-reflection-emit-ilgenerator-emitcalli) for reference.
+```diff
+namespace System.Reflection.Emit
+{
+    public class ILGenerator
+    {
++       public virtual void EmitCalli(
+            // Must be OpCodes.Calli which matches existing EmitCalli convention
+            // Not sure why this is redundant (for future Calli opcodes?)
++           System.Reflection.Emit.OpCodes opcode,
++           Type[] callingConventions,
++           Type? returnType,
++           Type[]? parameterTypes);
+
++       public virtual void EmitCalli(
++           System.Reflection.Emit.OpCodes opcode,
++           Type functionPointerType)
+    }
+}
+```
+
+# Breaking changes
+The breaking change is: **`typeof(delegate*<...>)` and `Type.GetType()` will return a `Type` instance instead of `IntPtr`**
```

Comment:
Observation: This breaking change is not actually strictly required to make the function pointers inspectable by reflection. We can keep the runtime current behavior and tell people to use signature types to gather any function pointer details.

(I am not trying to oppose the breaking change. Just sharing an observation.)

Reply:
Interesting; but I would say doing that goes against the goal of "correcting" the Type system.